### PR TITLE
[LightGBM] Implement `is_provide_training_metric` in Scala codes through JNI.

### PR DESCRIPTION
**Pull request by shishaochen**
In LightGBM CLI binary, there is an useful config parameter named `is_provide_training_metric`. If it is turned on, all defined metrics for training data are printed after each iteration like:
```
[LightGBM] [Info] Iteration:1, training auc : 0.973625
[LightGBM] [Info] Iteration:1, training binary_logloss : 0.252878
[LightGBM] [Info] 5.466509 seconds elapsed, finished iteration 1
[LightGBM] [Info] Iteration:2, training auc : 0.98312
[LightGBM] [Info] Iteration:2, training binary_logloss : 0.185529
[LightGBM] [Info] 11.769265 seconds elapsed, finished iteration 2
```
These messages are extremely helpful when drawing metric curves during training.

However, it won't work if we pass `is_provide_training_metric` to LightGBM as a config parameter. Because this parameter is only consumed in the [CLI version](https://github.com/microsoft/LightGBM/blob/master/src/application/application.cpp#L118). Thus, we can extract metrics in Scala codes through JNI and print them after each iteration.

With this pull request merged, training metrics will be printed in each Spark executor's STDERR like:
```
19/05/16 11:21:53 INFO LightGBMClassifier: LightGBM running iteration: 66 with result: 0 and is finished: false
19/05/16 11:21:53 INFO LightGBMClassifier: Train auc=0.988505429160646
19/05/16 11:21:53 INFO LightGBMClassifier: Train binary_logloss=0.10812842975017523
19/05/16 11:21:54 INFO LightGBMClassifier: Valid auc=0.9887551394869192
19/05/16 11:21:54 INFO LightGBMClassifier: Valid binary_logloss=0.1077190874299224
19/05/16 11:21:54 INFO LightGBMClassifier: LightGBM worker calling LGBM_BoosterUpdateOneIter
19/05/16 11:21:56 INFO LightGBMClassifier: LightGBM running iteration: 67 with result: 0 and is finished: false
19/05/16 11:21:56 INFO LightGBMClassifier: Train auc=0.9885683489789646
19/05/16 11:21:56 INFO LightGBMClassifier: Train binary_logloss=0.10773347478545063
19/05/16 11:21:57 INFO LightGBMClassifier: Valid auc=0.9888314443159278
19/05/16 11:21:57 INFO LightGBMClassifier: Valid binary_logloss=0.10733361894798527
```

### CODE REVIEW

**Comment by shishaochen** [comment]
Could you share error messages of this build?
At my DEV machine, I only built the part of LightGBM and it seemed no problem.

**Comment by shishaochen** [comment]
@mhamilton723 Could you help push forward this pull request? Thanks!
@imatiach-msft Kindly ping~

**Comment by mhamilton723** [comment]
Hey @shishaochen thanks for the ping. One of the services we tested against changed their outputs which caused a transient build failure. Could you please rebase on latest master and I will re-queue your build? Thanks!

**Comment by shishaochen** [comment]
@mhamilton723 My private branch is rebased onto master now. Thanks!

**Comment by imatiach-msft** [comment]
Seeing this error now:

##[error]$HOME/agent/_work/1/s/src/lightgbm/src/main/scala/TrainParams.scala:86:12: class RankerTrainParams needs to be abstract, since method isProvideTrainingMetric in class TrainParams of type => Boolean is not defined
##[error]case class RankerTrainParams(val parallelism: String, val numIterations: Int, val learningRate: Double,
##[error]^
##[error]one error found

**Comment by imatiach-msft** [comment]
Ah, it looks like you just need to add the param to LightGBMRanker which was added recently, similar to LightGBMClassifier and LightGBMRegressor.

**Comment by shishaochen** [comment]
@imatiach-msft The parameter `is_provide_training_metric` is added to LightGBMRanker in the new commit.
